pkg/dataconn/wire.go

```diff
@@ -6,39 +6,46 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"runtime"
 )
 
 type Wire struct {
-	conn   net.Conn
-	writer *bufio.Writer
-	reader io.Reader
+	conn      net.Conn
+	writer    *bufio.Writer
+	reader    io.Reader
+	byteOrder binary.ByteOrder
 }
 
 func NewWire(conn net.Conn) *Wire {
+	var byteOrder binary.ByteOrder = binary.LittleEndian
+	if runtime.GOARCH == "s390x" {
+		byteOrder = binary.BigEndian
+	}
```

Comment:
@keithalucas one thing to concern. I thought different architectures have different endian supports? If yes, probably we can't unify them? 

```
• Big-endian: A format in which the most significant byte is stored first. The other bytes follow in decreasing order of significance. The big-endian format is used by s390x (IBM Z & LinuxONE), Sun SPARC, and others.

• Little-endian: A format in which the least significant byte is stored first. The other bytes follow in increasing order of significance. The little-endian format is used by x64, x86, ARM, ARM64 among other processors.
```

`With this endianness change, we won't be able to use longhorn if some replicas are on a big endian system and some are on a little endian` <--- we only allow nodes running in the same architecture, so this can be defined.

Reply:
After discussing with @derekbit and @joshimoo , @keithalucas made the good point, and we should keep the consistent endian for the rx/tx on longhorn, but have the correct conversation at the end on the host arch.

@derekbit and @joshimoo will provide more feedbacks then.

Reply:
There are two parts of the endianness issue, one for data transmission and another for host.
We can choose either big endian or small endian for the transmission, but both the transmitter and receiver should follow the protocol. For example, if the transmitter sends the big-endian data, the receiver should know the endianness and convert the data to the host endian. The host endian can be also either big or small endian.

Reply:
@derekbit will the conversion cause extra costs impacting the performance?

Reply:
As per s390 will be experimental in 1.3.0, it seems fine for now w/ some constraints like unsupported hosts w/ different endianness. However, in the end, we should introduce the host endianness conversation w/ a consistent endianness used in Longhorn. WDYT? 

@derekbit @keithalucas @joshimoo ?

Reply:
If we stick with little endian for the protocol, there won't be any impact on the x86_64 or ARM.   The golang code without the s390 change handles little endian already.  For s390, liblonghorn would need to change to send the 16 and 32 bit unsigned ints as little endian.  Since liblonghorn is C, we could determine as compile time if the arch it little or big endian and use the C preprocessor to do nothing to convert to little endian on little endian architectures.

If we chose to always use big endian, there will be an impact on x86_64 or ARM but it should be minimal.   I thought it might make sense to use big endian because big endian is "Network byte order" which is often used in network protocols, but there are also network protocols that use little endian for transport on big endian systems, like protocol buffers.

Switching to big endian on x86_64 and ARM will affect interoperability between engines versions.  I believe even if the user chooses a custom engine image container for a volume, that engine instance will connect to the tgt/liblonghorn in the instance manager container.   Because of that, we should use little endian.

So I think we should choose an endianness for the protocol and not use the system's native endianness.    Also, I think lilttle endian should be used in the network messages.   The existing golang code is written to explicitly expect little endian.   The C code in liblonghorn uses whatever the system's endianness is.  liblonghorn should be changed to use little endian explicitly on big endian architectures.   We can determine the endianness when compiling liblonghorn by using something autoconf/automake or cmake.

Reply:
@innobead I can create the ticket.

Reply:
I agree with @keithalucas suggestion to stick with little endian in the go code. This way we stay backwards compatible and don't require changes here, since we explicitly specified little endian.

I further agree with the suggestion on making a compile time change on liblonghorn to specify the endianess for the transmission of the message fields.

This is a great example of a good PR review discussion
For knowledge sharing cc @shuo-wu @c3y1huang @PhanLe1010

Reply:
Based on the above the discussion, 
I updated the PR to remove the changes related to the little endian/big endian topic.

Many tests are failing at the moment on s390x but that's expected for now.

Reply:
After https://github.com/rancher/liblonghorn/pull/12 is merged we can update the ref on the engine to the new liblonghorn version. cc @innobead I do not have merge permissions on the liblonghorn repo, I did the review can you check if you can merge this pr.

ISSUE REF: https://github.com/longhorn/longhorn/issues/3875

Reply:
Merged and @keithalucas will update the lib here later.

Reply:
I updated lib longhorn in this PR.
Now all the tests are passing.